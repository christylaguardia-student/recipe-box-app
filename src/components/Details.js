import React from 'react';
import { fractions } from '../store/ingredient.constants.js';
import request from '../services/recipe-box.api';
import '../styles/Details.css';

export default function Details({ recipe }) {

  return (
    <div>
      <h1>{recipe.title}</h1>
      <button onClick={() => {alert("this does nothing")} }>Edit</button>
      <button onClick={() => remove(recipe._id) }>Delete</button>

      {recipe.servings ? <p>Serves: {recipe.servings}</p> : null }
      {recipe.time ? <p>Time: {convertTime(recipe.time)}</p> : null }

      <h3>Ingredients</h3>
      <ul>
        {recipe.ingredients.map((item, index) => {
          return <li key={item._id}>
            <input type="checkbox" />{convertToFraction(item.amount)} {item.unit} {item.name}
            </li>
        })}
      </ul>

      <h3>Instructions</h3>
      <p>{recipe.instructions}</p>
    </div>
  )

}

function remove(id) {
  request.delete(id)
    .then(removed => {
      // TODO: show msg
    })
    .catch((err) => {
      console.log('there was a problem deleing that recipe', err);
    })
}

function convertTime(time) {
  if (!time) return '?';

  const hours = Math.floor(time / 60);
  const minutes = time % 60;

  if ((time / 60) > 1) return `${hours} hours and ${minutes} minutes`;
  else return `${minutes} minutes`;
}

function convertToFraction(amount) {
  const hasFraction = amount.toString().includes('.');

  if (hasFraction) {
    const nums = amount.toString().split('.');
    const whole = parseInt(nums[0], 10);
    const decimal = parseFloat(`.${nums[1]}`);
    const fraction = fractions.find(f => f.decimal === decimal);

    if (whole > 0) amount = `${whole} ${fraction.value}`;
    else amount = fraction.value;
  }

  return amount;
}

// function scale(ingredients, factor) {

//   let scaledIngredients = ingredients.map(item => {
//     const newAmount = item.amount * factor;

//     return {
//       _id: item._id,
//       name: item.name,
//       amount: newAmount,
//       unit: item.unit
//     };

//   });

//   console.log(this.state.ingredients, scaledIngredients);
// }
