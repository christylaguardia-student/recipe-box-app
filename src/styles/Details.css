
/* #recipe-details {
  margin-left: 30vh;
} */

ul {
  list-style-type: none;
}