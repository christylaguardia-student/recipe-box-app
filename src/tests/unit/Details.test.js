import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import Details from '../../components/Details';

describe('<Details />', () => {

  const testRecipe = {
    "_id": 1,
    "title": "Cookies",
    "instructions": "Mix stuff and put on cookie sheet. Then put in oven. Do not burn.",
    "ingredients": [ 
      { "_id": 1, "name": "flour", "amount": 2, "unit": "cup" },
      { "_id": 2, "name": "sugar", "amount": 1, "unit": "cup" },
      { "_id": 3, "name": "chocolate chips", "amount": 2.5, "unit": "cup" }
    ]
  }

  it('renders the Details component', () => {
    const wrapper = toJSON(shallow(<Details recipe={testRecipe} />));
    expect(wrapper).toMatchSnapshot();
  });

})