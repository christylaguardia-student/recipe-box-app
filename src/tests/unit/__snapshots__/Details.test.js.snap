// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Details /> renders the Details component 1`] = `
<div>
  <h2>
    Cookies
  </h2>
  <p>
    Serves: 
  </p>
  <p>
    Time: 
    ?
  </p>
  <h3>
    Ingredients
  </h3>
  <ul>
    <li>
      <input
        type="checkbox"
      />
       
      2
       
      cup
       
      flour
    </li>
    <li>
      <input
        type="checkbox"
      />
       
      1
       
      cup
       
      sugar
    </li>
    <li>
      <input
        type="checkbox"
      />
       
      2 1/2
       
      cup
       
      chocolate chips
    </li>
  </ul>
  <h3>
    Instructions
  </h3>
  <p>
    Mix stuff and put on cookie sheet. Then put in oven. Do not burn.
  </p>
  <button
    onClick={[Function]}
  >
    Edit
  </button>
</div>
`;
